doc/keynames/keynames.py

```diff
@@ -0,0 +1,470 @@
+#!/bin/python3
```

Comment:
As explained, in the accompanying README, this is not meant as a 100% correct reference, but more as a very likely to be correct implementation guide.

I created it, when I found a bug in the C code to help me debug what was happening and what should actually happen. I think this Python implementation is much easier to understand than the C version (which tries to do as much as possible in a single iteration of the string).

I have not finished the README yet, but AFAIK right now, this Python code actually defines how Key Names work. I will also extend the CLI, so that it can generate test cases for `test_keyname.c`.

If anything about Key Names changes in the future, the procedure would be:
1. Update the Python code, to make sure your changes actually make sense.
2. Update the documentation (README.md)
3. Use the Python CLI to generate new test cases.
4. Update the C code to match the changes.

Reply:
Thank you :sparkling_heart:  for this reference implementation, this totally makes sense and will help @PhilippGackstatter a lot, as the C implementation is scattered around and more complicated (and might get even more complicated when we start to optimize it).

I would, however, prefer that doc/keynames/README.md  goes to doc/dev/keynames.md and this file goes to scripts/dev/keynames.py. As the md refers to the py, it should be easy to find.

Reply:
>  as the C implementation is scattered around and more complicated (and might get even more complicated when we start to optimize it)

It is actually contained in just a few functions as well:
- `elektraKeyNameValidate`: checks if a string is a valid key name and also calculates the size of both the canonical escaped and the unescaped names. This already done in (mostly) a single iteration of the string. As far as possible it examines each character just once. This by far the most complicated of the functions because it does a lot of things simultaneously and because of the `..` (specifically for the `keyAddName` case) has to check the whole key name in reverse.
- `elektraKeyNameCanonicalize`: does an iteration similar to `elektraKeyNameValidate` and creates a canonical key name with about one `memcpy` per key name part. In some cases fewer `memcpy` might be possible, but because of `.`, `..` and array parts it is very hard to detect such cases.
- `elektraKeyNameUnescape`: takes the canonical name and unescapes it. It already tries to use as few `memcpy` as possible, i.e. one base plus just one per `%` or `\`. If the canonical name contains neither `%` nor `\`, the function is just a `strpbrk`, a `memcpy` and an iteration of the key in which we replace `/` with `\0`.

There might be potential for optimization, but it would require **very** extensive benchmarking and will probably involve some `#ifdef` and compiler builtins. At least with the current rules I don't see much potential.

> I would, however, prefer that doc/keynames/README.md goes to doc/dev/keynames.md and this file goes to scripts/dev/keynames.py. As the md refers to the py, it should be easy to find.

:+1: I'll move it over there at the end of the PR. Moving now might make GitHub loose the comments.

Reply:
Yes, I will add it. But probably not to the markdown file, since that has a different target audience. I will add it to the `@defgroup` doxygen comment at the top of `keyname.c`.